### PR TITLE
[LoongArch] Use R_LARCH_ALIGN without symbol as much as possible

### PR DESCRIPTION
**Pull request by SixWeining**
To support the third parameter of the alignment directive, R_LARCH_ALIGN relocations need a non-zero symbol index.
In many cases we don't need the third parameter and can set the symbol index to 0.
This patch will remove a lot of .Lla-relax-align* symbols and mitigate the size regression due to https://github.com/llvm/llvm-project/pull/72962.

This patch is mainly from He Jinyang <hejinyang@loongson.cn>, also of Loongson.

Co-authored-by: Weining Lu <luweining@loongson.cn>

### CODE REVIEW

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp [hunk 1]
```diff
@@ -224,20 +224,25 @@ bool LoongArchAsmBackend::shouldInsertFixupForCodeAlign(
...
-    Sym->setFragment(&*Sec->getBeginSymbol()->getFragment());
-    Asm.registerSymbol(*Sym);
-    MCSym = MCSymbolRefExpr::create(Sym, Ctx);
-    getSecToAlignSym()[Sec] = MCSym;
-  }
+
+  auto createExtendedValue = [&]() {
+    const MCSymbolRefExpr *MCSym = getSecToAlignSym()[Sec];
+    if (MCSym == nullptr) {
+      // Create a symbol and make the value of symbol is zero.
```
Comment (MaskRay) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:231]: Define a marker symbol at the section with an offset of 0.

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp [hunk 2]
```diff
@@ -224,20 +224,25 @@ bool LoongArchAsmBackend::shouldInsertFixupForCodeAlign(
...
+      MCSymbol *Sym = Ctx.createNamedTempSymbol("la-relax-align");
+      Sym->setFragment(&*Sec->getBeginSymbol()->getFragment());
+      Asm.registerSymbol(*Sym);
+      MCSym = MCSymbolRefExpr::create(Sym, Ctx);
+      getSecToAlignSym()[Sec] = MCSym;
+    }
+    unsigned Lo = Log2_64(Count) + 1;
```
Comment (MaskRay) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:238]: prefer `countr_zero` from llvm/ADT/bit.h

**Comment by MaskRay** [comment]
"Squash and merge" assigns the author field to the PR author (not the patch author) and adds Co-authored-by tags for commit authors.
If you are a coauthor, you can set the Co-authored-by tag in the description.

> In LoongArch psABI v2.30, the R_LARCH_ALIGN requires symbol index to to support the third parameter of alignment directive. Following this extended R_LARCH_ALIGN make the symbol index be redundant in some case. E.g. .align 4 or .align 4,,16. Thus, use the extended R_LARCH_ALIGN only when the third parameter is effective. Otherwise use the orginal R_LARCH_ALIGN for simple.

"... to support the third parameter of the alignment directive, R_LARCH_ALIGN relocations need a non-zero symbol index.
In many cases we don't need the third parameter  and can set the symbol index to 0.
This patch will remove a lot of `.Lla-relax-align*` symbols and mitigate the size regression due to #72962."

**Review by MaskRay — APPROVED** [review]

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp (hunk 2 above)
Comment (SixWeining) [reply llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:238]: Thanks, but I find `Log2(AF.getAlignment())` is simpler.

llvm/Support/Alignment.h:
```
/// Returns the log2 of the alignment.
inline unsigned Log2(Align A) { return A.ShiftValue; }
```

**Pull request merged**